flang/lib/Optimizer/Builder/IntrinsicCall.cpp

```diff
@@ -8060,6 +8061,21 @@ mlir::Value IntrinsicLibrary::genSind(mlir::Type resultType,
   return getRuntimeCallGenerator("sin", ftype)(builder, loc, {arg});
 }
 
+// SINPI
+mlir::Value IntrinsicLibrary::genSinpi(mlir::Type resultType,
+                                       llvm::ArrayRef<mlir::Value> args) {
+  assert(args.size() == 1);
+  mlir::MLIRContext *context = builder.getContext();
+  mlir::FunctionType ftype =
+      mlir::FunctionType::get(context, {resultType}, {args[0].getType()});
+  llvm::APFloat pi = llvm::APFloat(llvm::numbers::pi);
+  mlir::Value dfactor =
+      builder.createRealConstant(loc, mlir::Float64Type::get(context), pi);
+  mlir::Value factor = builder.createConvert(loc, args[0].getType(), dfactor);
```

Comment:
Why not create the real constant with the argument type in the first place? This might matter for real(16).

Reply:
Ahh okay what I think is happening here is that APFloat is forcing its size to f64 because `llvm::numbers::pi` is defined as a double. I'm concerned that this could produce inaccurate results for real(16) (aka floating point types with more precision than `f64`: f128 on aarch64 systems, I think f80 on x86).

Reply:
As I understand it, flang-rt is set up to call either glibc or libquadmath depending what is available. If possible it would be good to re-use that mechanism instead of adding further logic for that here. So far as I can tell, the runtime call generator is calling the right runtime function for f128, at least for sind.

But yes I see what you mean about the 64-bit constants used with sind. I think merging this is okay so long as this does get fixed later.

Reply:
Diverge in what sense?

If we "lower" the f32/f64 cases here as `sin(PI * x)`, and the f128 case in Flang's `quadmath` as `sinq(M_PIq * x)`, then they should be pretty consistent, right?